contracts/carrot-app/src/handlers/execute.rs

```diff
@@ -240,10 +248,40 @@ fn _inner_withdraw(
         .transpose()?;
 
     // We withdraw the necessary share from all registered investments
-    let withdraw_msgs =
-        STRATEGY_CONFIG
-            .load(deps.storage)?
-            .withdraw(deps.as_ref(), withdraw_share, app)?;
+    let mut withdraw_msgs: Vec<CosmosMsg> = STRATEGY_CONFIG
+        .load(deps.storage)?
+        .withdraw(deps.as_ref(), withdraw_share, app)?
+        .into_iter()
+        .map(Into::into)
+        .collect();
+
+    if let Some(swap_to) = swap_to {
+        let withdraw_preview = STRATEGY_CONFIG.load(deps.storage)?.withdraw_preview(
+            deps.as_ref(),
+            withdraw_share,
+            app,
+        )?;
+
+        // We swap all withdraw_preview coins to the swap asset
+        let config = CONFIG.load(deps.storage)?;
+        let ans = app.name_service(deps.as_ref());
+        let dex = app.ans_dex(deps.as_ref(), config.dex);
+        withdraw_preview.into_iter().try_for_each(|fund| {
+            let asset = ans.query(&Asset::from(fund.clone()))?;
+            if asset.name != swap_to {
+                let swap_msg = dex.swap(
+                    asset,
+                    swap_to.clone(),
+                    Some(MAX_SPREAD),
+                    Some(query_exchange_rate(deps.as_ref(), fund.denom, app)?),
+                )?;
+                withdraw_msgs.push(swap_msg);
+            }
+            Ok::<_, AppError>(())
+        })?;
+    }
+
+    deps.api.debug(&format!("{:?}", withdraw_msgs));
```

Comment:
about gas situation, noticed you have a lot of debugs. Want to mention that it will still eat gas for converting variable to debug string, and to allocate formatted string even if it does not get printed. Not really sure how much it is actually, but just to keep in mind before next gas measurements

Reply:
Oh yes you are right ! This is very low though compared to what the app is actually eating !